crates/provider/src/layers/batch.rs

```diff
@@ -0,0 +1,515 @@
+use crate::{
+    bindings::IMulticall3, Caller, Provider, ProviderCall, ProviderLayer, RootProvider,
+    MULTICALL3_ADDRESS,
+};
+use alloy_eips::BlockId;
+use alloy_network::{Ethereum, Network, TransactionBuilder};
+use alloy_primitives::{Address, Bytes, U256};
+use alloy_rpc_client::WeakClient;
+use alloy_sol_types::{SolCall, SolType, SolValue};
+use alloy_transport::{utils::Spawnable, TransportErrorKind, TransportResult};
+use std::{fmt, future::IntoFuture, marker::PhantomData, sync::Arc, time::Duration};
+use tokio::sync::{mpsc, oneshot};
+
+#[cfg(target_arch = "wasm32")]
+use wasmtimer::tokio::sleep;
+
+#[cfg(not(target_arch = "wasm32"))]
+use tokio::time::sleep;
+
+const DEFAULT_WAIT: Duration = Duration::from_millis(1);
+
+/// Provider layer that aggregates contract calls (`eth_call`) over a time period into a single
+/// [Multicall3] contract call.
+///
+/// Some methods, such as `eth_getBlockNumber`, are first converted into contract calls to the
+/// [Multicall3] contract itself and then aggregated with other `eth_call`s.
+///
+/// Only calls that:
+/// - target the latest block ID,
+/// - have no state overrides,
+/// - have a target address and calldata,
+/// - have no other properties (nonce, gas, etc.)
+///
+/// can be sent with a multicall.
+///
+/// This layer is useful for reducing the number of network requests made.
+/// However, this only works when requests are made in parallel, for example when using the
+/// [`tokio::join!`] macro or in multiple threads/tasks, as otherwise the requests will be sent one
+/// by one as normal, but with an added delay.
+///
+/// [Multicall3]: https://github.com/mds1/multicall3
+#[derive(Debug)]
+pub struct BatchLayer {
+    m3a: Address,
+    wait: Duration,
+}
+
+impl Default for BatchLayer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl BatchLayer {
+    /// Create a new `BatchLayer` with a default wait of 1ms.
+    pub fn new() -> Self {
+        Self { m3a: MULTICALL3_ADDRESS, wait: DEFAULT_WAIT }
+    }
+
+    /// Set the amount of time to wait before sending the batch.
```

Comment:
can we clarify here that this is the time it waits for additional calls before sending the the call, so this the max delay overhead per call